File: android/app/src/main/java/chat/ssh/sshchat/SecureDownload.kt
package chat.ssh.sshchat

import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

data class DownloadedMedia(
    val file: File,
    val name: String,
    val mime: String,
    val sender: String? = null,
) {
    val isImage: Boolean
        get() = MediaMime.isImage(mime)
    val isVideo: Boolean
        get() = MediaMime.isVideo(mime)
    val isAudio: Boolean
        get() = MediaMime.isAudio(mime)
}

object SecureDownload {
    fun fetch(pageUrl: String, key: String, outDir: File): DownloadedMedia {
        val (base, token) = splitDownloadUrl(pageUrl)
        outDir.mkdirs()
        val ticketBody = JSONObject().put("key", key.uppercase()).toString()
        val ticketJson = httpJson(
            "$base/download/$token/ticket",
            method = "POST",
            body = ticketBody,
            contentType = "application/json",
        )
        val filename = ticketJson.optString("filename").ifBlank { "file" }
        val mime = ticketJson.optString("mime").ifBlank {
            MediaMime.guess(filename)
        }
        val rel = ticketJson.optString("download")
        if (rel.isBlank()) error("no download ticket")
        val fileUrl = URL(URL(base + "/"), rel.removePrefix("/")).toString()
        val bytes = httpBytes(fileUrl)
        val safe = filename.replace(Regex("""[\\/]"""), "_").take(180).ifBlank { "file" }
        val out = File(outDir, "${System.currentTimeMillis()}-$safe")
        out.writeBytes(bytes)
        return DownloadedMedia(out, filename, mime)
    }

    private fun splitDownloadUrl(url: String): Pair<String, String> {
        val u = URL(url)
        val parts = u.path.split('/').filter { it.isNotEmpty() }
        require(parts.size >= 2 && parts[0] == "download") { "invalid download url" }
        val base = "${u.protocol}://${u.authority}"
        return base to parts[1]
    }

    private fun httpJson(
        url: String,
        method: String,
        body: String? = null,
        contentType: String? = null,
    ): JSONObject {
        val raw = String(httpBytes(url, method, body?.toByteArray(StandardCharsets.UTF_8), contentType), StandardCharsets.UTF_8)
        return JSONObject(raw.ifBlank { "{}" })
    }

    private fun httpBytes(
        url: String,
        method: String = "GET",
        body: ByteArray? = null,
        contentType: String? = null,
    ): ByteArray {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 20_000
            readTimeout = 180_000
            doInput = true
            if (body != null) {
                doOutput = true
                if (contentType != null) setRequestProperty("Content-Type", contentType)
                outputStream.use { it.write(body) }
            }
        }
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val bytes = stream?.use { it.readBytes() } ?: ByteArray(0)
        if (code !in 200..299) {
            val msg = runCatching {
                JSONObject(String(bytes, StandardCharsets.UTF_8)).optString("error")
            }.getOrNull()?.takeIf { it.isNotBlank() }
                ?: String(bytes, StandardCharsets.UTF_8).take(200).ifBlank { "HTTP $code" }
            error(msg)
        }
        return bytes
    }
}

File: ios/SSHChat/SecureDownload.swift
import Foundation

enum SecureDownload {
    static func fetch(pageURL: String, key: String, outDir: URL) async throws -> DownloadedMedia {
        let (base, token) = try splitDownloadURL(pageURL)
        try FileManager.default.createDirectory(at: outDir, withIntermediateDirectories: true)

        let ticketURL = URL(string: "\(base)/download/\(token)/ticket")!
        var ticketReq = URLRequest(url: ticketURL)
        ticketReq.httpMethod = "POST"
        ticketReq.timeoutInterval = 180
        ticketReq.setValue("application/json", forHTTPHeaderField: "Content-Type")
        ticketReq.httpBody = try JSONSerialization.data(withJSONObject: ["key": key.uppercased()])

        let (ticketData, ticketResp) = try await URLSession.shared.data(for: ticketReq)
        try throwIfBad(ticketResp, ticketData)
        let ticket = (try? JSONSerialization.jsonObject(with: ticketData) as? [String: Any]) ?? [:]
        let filename = (ticket["filename"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let name = (filename?.isEmpty == false) ? filename! : "file"
        var mime = (ticket["mime"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if mime.isEmpty { mime = MediaMime.guess(name) }
        guard let rel = ticket["download"] as? String, !rel.isEmpty else {
            throw DownloadError.message("no download ticket")
        }
        let fileURL = URL(string: rel.hasPrefix("/") ? "\(base)\(rel)" : "\(base)/\(rel)")!
        let (bytes, fileResp) = try await URLSession.shared.data(from: fileURL)
        try throwIfBad(fileResp, bytes)

        let safe = name
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
        let clipped = String(safe.prefix(180)).isEmpty ? "file" : String(safe.prefix(180))
        let out = outDir.appendingPathComponent("\(Int(Date().timeIntervalSince1970 * 1000))-\(clipped)")
        try bytes.write(to: out)
        return DownloadedMedia(url: out, name: name, mime: mime)
    }

    private static func splitDownloadURL(_ url: String) throws -> (String, String) {
        guard let u = URL(string: url) else { throw DownloadError.message("invalid download url") }
        let parts = u.path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        guard parts.count >= 2, parts[0] == "download" else {
            throw DownloadError.message("invalid download url")
        }
        let base = "\(u.scheme ?? "https")://\(u.host ?? "")\(u.port.map { ":\($0)" } ?? "")"
        return (base, parts[1])
    }

    private static func throwIfBad(_ resp: URLResponse, _ data: Data) throws {
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if (200...299).contains(code) { return }
        let raw = String(data: data, encoding: .utf8) ?? ""
        let json = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
        let err = (json["error"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        throw DownloadError.message(err?.isEmpty == false ? err! : (raw.prefix(200).isEmpty ? "HTTP \(code)" : String(raw.prefix(200))))
    }

    enum DownloadError: LocalizedError {
        case message(String)
        var errorDescription: String? {
            switch self {
            case .message(let s): return s
            }
        }
    }
}
